// @admin-bro/design-system/build/atoms/table/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var table_1 = require("./table");
Object.defineProperty(exports, "Table", { enumerable: true, get: function () { return table_1.default; } });
var table_head_1 = require("./table-head");
Object.defineProperty(exports, "TableHead", { enumerable: true, get: function () { return table_head_1.default; } });
var table_body_1 = require("./table-body");
Object.defineProperty(exports, "TableBody", { enumerable: true, get: function () { return table_body_1.default; } });
var table_cell_1 = require("./table-cell");
Object.defineProperty(exports, "TableCell", { enumerable: true, get: function () { return table_cell_1.default; } });
var table_row_1 = require("./table-row");
Object.defineProperty(exports, "TableRow", { enumerable: true, get: function () { return table_row_1.default; } });
var table_caption_1 = require("./table-caption");
Object.defineProperty(exports, "TableCaption", { enumerable: true, get: function () { return table_caption_1.default; } });
