// @admin-bro/nestjs/build/admin.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var AdminModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdminModule = void 0;
const admin_bro_1 = __importDefault(require("admin-bro"));
const common_1 = require("@nestjs/common");
const core_1 = require("@nestjs/core");
const serve_static_provider_1 = require("./serve-static.provider");
const token_constants_1 = require("./token.constants");
const abstract_loader_1 = require("./loaders/abstract.loader");
let AdminModule = AdminModule_1 = class AdminModule {
    constructor(httpAdapterHost, loader, adminModuleOptions) {
        this.httpAdapterHost = httpAdapterHost;
        this.loader = loader;
        this.adminModuleOptions = adminModuleOptions;
    }
    static createAdmin(options) {
        return {
            module: AdminModule_1,
            providers: [
                {
                    provide: token_constants_1.CONFIG_TOKEN,
                    useValue: options,
                },
            ],
        };
    }
    static createAdminAsync(options) {
        return {
            imports: options.imports,
            module: AdminModule_1,
            providers: [
                {
                    provide: token_constants_1.CONFIG_TOKEN,
                    useFactory: options.useFactory,
                    inject: options.inject,
                },
            ],
        };
    }
    onModuleInit() {
        const admin = new admin_bro_1.default(this.adminModuleOptions.adminBroOptions);
        const { httpAdapter } = this.httpAdapterHost;
        this.loader.register(admin, httpAdapter, Object.assign(Object.assign({}, this.adminModuleOptions), { adminBroOptions: admin.options }));
    }
};
AdminModule = AdminModule_1 = __decorate([
    common_1.Module({
        providers: [serve_static_provider_1.serveStaticProvider],
    }),
    __param(2, common_1.Inject(token_constants_1.CONFIG_TOKEN)),
    __metadata("design:paramtypes", [core_1.HttpAdapterHost,
        abstract_loader_1.AbstractLoader, Object])
], AdminModule);
exports.AdminModule = AdminModule;
